File: src/common/py/core/messaging/command.py
import typing
import uuid
from dataclasses import dataclass, field

from .message import Message, Trace


@dataclass(frozen=True, kw_only=True)
class Command(Message):
    """
    A command message.
    
    Commands are instructions that need to be 'executed' by the receiving component.
    
    Notes:
          Commands need to `always` be replied by emitting a corresponding :class:`CommandReply`.
          This reply is then automatically sent back to the original sender.
    
    Attributes:
        unique: A unique identifier for each issued command.
    """
    unique: Trace = field(default_factory=uuid.uuid4)


CommandType = typing.TypeVar("CommandType", bound=Command)

File: src/common/web/core/messaging/Command.ts
// @ts-ignore
import { v4 as uuidv4 } from "uuid";
import { Message, type MessageCategory, type Trace } from "./Message";

/**
 * A command message.
 *
 * Commands are instructions that need to be *executed* by the receiving component.
 *
 * Notes:
 *     Commands need to *always* be replied by emitting a corresponding ``CommandReply``.
 *     This reply is then automatically sent back to the original sender.
 */
export class Command extends Message {
    public static readonly Category: MessageCategory = "Command";

    public readonly unique: Trace = uuidv4();

    /**
     * Gets the global message category.
     */
    public get messageCategory(): MessageCategory {
        return Command.Category;
    }
}
